File: Chapter.14/StockTicker.Server.CSharp/Controllers/TradingController.cs
using System;
using System.Reactive.Subjects;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.SignalR;
using StockTicker.Core;
using static StockTicker.Core.Models;

namespace StockTicker.Server.CSharp.Controllers
{
    public class TradingController : Controller, IObservable<CommandWrapper>
    {
        private readonly IHubContext<Core.StockTicker> _hubContext;
        private readonly Subject<CommandWrapper> _subject;

        public TradingController(IHubContext<Core.StockTicker> hubContext)
        {
            _hubContext = hubContext;
            _subject = new Subject<CommandWrapper>();
        }


        public IDisposable Subscribe(IObserver<CommandWrapper> observer)
        {
            return _subject.Subscribe(observer);
        }

        private IActionResult Publish(string connectionId, Result<ClientOrder> cmd)
        {
            var hasObs = _subject.HasObservers;
            if (cmd.IsOk && hasObs)
            {
                var wrapped = CommandWrapper.CreateTrading(connectionId, cmd.Ok);
                _subject.OnNext(wrapped);
                return StatusCode(200);
            }

            _subject.OnError(cmd.Error);
            return StatusCode(500);
        }

        private IActionResult ProcessRequest(string connId, ClientOrder order)
        {
            var result = Validation.tradingdValidation.Invoke(order).ToResult();
            return Publish(connId, result);
        }

        public IActionResult Index()
        {
            return View("Index");
        }


        [HttpPost]
        [Route("trading/logportfolio")]
        public ActionResult<User> LogPortfolio([FromForm] User userMarket)
        {
            var userName = userMarket.Username;
            var initialCash = userMarket.Initialcash;
            return RedirectToAction(nameof(OpenDashboard), new {username = userName, initialcash = initialCash});
        }

        [HttpGet]
        [Route("trading/opendashboard")]
        public ActionResult<User> OpenDashboard(string userName, decimal initialCash)
        {
            var userModel = new User
                {Username = userName, Initialcash = initialCash, UserId = Guid.NewGuid().ToString("N")};

            // Load stocks by the username 
            return View("Market", userModel);
        }


        [Route("trading/placeorder")]
        [HttpPost]
        public IActionResult PlaceOrder([FromBody] Order data)
        {
            if (data == null)
                return BadRequest("POST body is null");

            var tradingType =
                string.Compare("buy", data.TradingType, StringComparison.OrdinalIgnoreCase) == 0
                    ? TradingType.Buy
                    : TradingType.Sell;

            var clientOrder = new ClientOrder
            {
                Price = decimal.Parse(data.Price),
                Quantity = data.Quantity,
                Symbol = data.Symbol.ToUpper(),
                TradingType = tradingType,
                UserId = data.UserId
            };

            return ProcessRequest(data.ConnId, clientOrder);
        }

        protected void Dispose()
        {
            _subject.Dispose();
        }
    }
}

File: Chapter.14/StockTicker.Server.FSharp/Controllers/TradingController.fs
namespace StockTicker.Server.FSharp.Controllers

open System
open System.Reactive.Subjects
open StockTicker.Core
open StockTicker.Core.Logging
open StockTicker.Core.Logging.Message
open Microsoft.AspNetCore.Mvc
open Microsoft.AspNetCore.SignalR
open StockTicker.Server.FSharp

//Listing 14.1 Web-API trading controller
type TradingController(hubContext: IHubContext<StockTickerHub.StockTicker>) =
    inherit Controller()

    static let logger = Log.create "StockTicker.TradingController"
    let logUserRequest methodName (tr : Order) =
        logger.logWithAck Info (
           eventX "{logger}: Called {url} with trading request {tr}"
           >> setField "logger" (sprintf "%A" logger.name)
           >> setField "url" ("/api/trading/"+methodName)
           >> setField "tr" ((sprintf "%A" tr).Replace("\n","")) )

    // the controller act as a observable publisher of messages
    // keep the controller loosely coupled using the Reactive Extensions
    // defining a Reactive Subject which will fire up
    // requests to the subscriber(s)
    let subject = new Subject<CommandWrapper>()  // #A

    let log res =
        logger.logWithAck Info (
            match res with
            | Result.Ok m ->
                eventX "{logger}: Validation successful - {req}"
                >> setField "req" ((sprintf "%A" m).Replace("\n",""))
            | Result.Error f ->
                eventX "{logger}: Validation failed - {msg}"
                >> setField "msg" f
            >> setField "logger" (sprintf "%A" logger.name))
        |> Async.RunSynchronously
        res

    let publish connectionId cmd =   // #G
        if subject.HasObservers then 
            match cmd with
            | Result.Ok(cmd) ->         // #B
                CommandWrapper.CreateTrading connectionId cmd// #C
                |> subject.OnNext       // #A
            | Result.Error(e) -> subject.OnError(exn (e))    // #B
        cmd
        
    let processRequest (connId: string) (order: ClientOrder) =
        Validation.tradingdValidation order
        |> log
        |> publish connId 
                
    [<HttpGet>]
    member this.Index() : IActionResult =
        base.View("Index") :> IActionResult
        
    [<HttpPost>]
    [<Route("trading/logportfolio")>]
    member this.LogPortfolio([<FromForm>]userMarket: User) =       
        let userName = userMarket.Username;
        let initialCash = userMarket.Initialcash;
        base.RedirectToAction("OpenDashboard",  {| userName = userName; initialCash = initialCash |} )
       
    [<HttpGet>]
    [<Route("trading/opendashboard")>]
    member this.OpenDashboard(userName: string, initialCash: decimal) =
        let userModel = 
            { User.Username = userName; Initialcash = initialCash; UserId = Guid.NewGuid().ToString("N") }

        // Load stocks by the username 
        base.View("Market", userModel)


    member private this.toResponse result =
        let response =
            match result with
            | Ok(_) -> base.StatusCode(200)
            | _ -> base.StatusCode(500)    // #D
        response            
    
    [<Route("trading/placeorder")>]
    [<HttpPost>]
    member this.PlaceOrder([<FromBody>]data: Order) =
        // can easily make asynchronous controller methods.
        Async.StartAsTask <| async {        // #H
        
        do! logUserRequest data.TradingType data
        let connectionId = data.ConnId
        
        let tradingType =
            if String.Compare("buy", data.TradingType, StringComparison.OrdinalIgnoreCase) = 0 then
                TradingType.Buy
            else TradingType.Sell;
            
        return {
            ClientOrder.Price = Decimal.Parse(data.Price)
            Quantity = data.Quantity
            Symbol = data.Symbol
            TradingType = tradingType
            UserId = data.UserId
        }
        // validation using function composition
        |> tradingdValidation           // #F
        |> log
        |> publish connectionId         // #G
        |> this.toResponse              // #D
    }
        
    // The controller behaves as Observable publisher and it can be register
    interface IObservable<CommandWrapper> with
        member this.Subscribe observer = subject.Subscribe observer  // #A

    override this.Dispose disposing =
        if disposing then subject.Dispose()
        base.Dispose disposing    // #I

